### ISSUE TITLE
Instructions on how to build locally are not fully correct

### ISSUE THREAD

**Issue by baumgarb** [dynatrace-oss/dynatrace-monitoring-as-code#37]
The CONTRIBUTING.MD file says that `go build ./...` should work fine but it seems that this is not correct. Instead, being in the project's root directory you need to run `go build ./cmd/...` it seems and then it spits out the `monaco` binary. 

If I'm not overlooking anything then this should be fixed in the documentation. If no one has any more input on this I'd gladly adjust the documentation and create a PR for it :)

**Comment by UnseenWizzard** [comment]
Yep, we apparent missed that when moving to more nested structure of packages & cmd

> When compiling multiple packages or a single non-main package, build compiles the packages but discards the resulting object, serving only as a check that the packages can be built.

I think the documentation should simply be changed to reference the `Makefile` rather than the exact go commands. 

If we add e.g. a second binary in the future that will need to be explicitly built as well - which is an easy extra step in the make, without having to change workflow otherwise.

**Comment by baumgarb** [comment]
Hm I'm not entirely sure because the CONTRIBUTING.MD mentions a lot of go commands (install, build OS specific, test, fmt etc.). Should we replace all these mentions of a go command with a corresponding make file invocation? We would probably have to modify the `Makefile` to support OS specific builds for example as well, don't we?

And also, to support x-platform development, the `Makefile` as it is right now does not work on Windows out of the box, does it?  The go commands on the other hand work the same way on Linux and Windows.

**Comment by UnseenWizzard** [comment]
Fair point. 
Seems we've dropped the `release` step that built different platforms from the Makefile. 

I think the steps of the make file that won't work x-platform/on Windows are `lint` and `clean` right?
For lint there is sadly no out of the box way to get an error-code for wrong formatting from go - thus the shell script turning `gofmt`   return values into an error or OK result. 

I generally agree that having the base Go commands is the best way to go, thus they're still in the doc so far. 
What I dislike is that they've already run out of date to how we actually build since the introduction of the Makefile. 

Maybe we should list the make steps and point people to look at the Makefile for the actual go commands? 
And introduce steps for `go mod tidy` (could be part of format as well), `go generate` (could be part of test or build)? 

x-platform executables can already be built with the make step (just tested), we'll just have to document that
e.g.  `make build GOOS=windows GOARCH=386 BINARY=monaco-win.exe`

**Comment by UnseenWizzard** [comment]
@patrickpichler what's your opinion on this?

**Comment by patrickpichler** [comment]
I personally prefer having a Makefile. If you ask me it is easier and less error prone
to simply run `make` instead of `go build ./cmd/...`. If for whatever reason you need
some special flags, you can always execute the cmd directly, but for most cases 
the make file command fits. 

Another advantage of using make is, that in the CI we also simply call make. So 
if we would need to change the way we build, we only would need to update the
make file.

**Comment by patrickpichler** [comment]
As for OS specific commands like `clean`, `lint`: We could simply include an `if` in the make goal. 
So if we detect you are running windows, we simply use the windows equivalent of `rm`.

**Comment by UnseenWizzard** [comment]
https://github.com/dynatrace-oss/dynatrace-monitoring-as-code/pull/72 makes the Make targets work on Windows & updates CONTRIBUTING.md

**Comment by baumgarb** [comment]
Quick question since I'm not too familiar with Go and with any other open source Go projects: what are other open source Go projects using? A `make` file or `go ...` commands or both or shell scripts and batch files or ... :-) I guess the best option is to adhere to best practises out there and I'd assume there are some in this specific area.

**Comment by patrickpichler** [comment]
Here are a few projects which use Makefile:

- https://git.sr.ht/~sircmpwn/aerc 
- https://github.com/kubernetes/kubernetes
- https://github.com/go-jira/jira
- https://github.com/gopasspw/gopass

To be honest almost all open source go projects i know of use Makefiles ^^

**Comment by 163a** [comment]
I am not a developer, but when I see a Makefile, I intuitively know to run `make clean && make build`
The tricky part was to understand why OneAgent breaks this.

```
/opt/repos/dynatrace-monitoring-as-code$ make clean && make build
Remove monaco and bin/
Remove monaco and bin/
Build monaco
ERROR: ld.so: object '/lib/x86_64-linux-gnu/liboneagentproc.so' from /etc/ld.so.preload cannot be preloaded (cannot open shared object file): ignored.
go: downloading github.com/golang/mock v1.4.4
go: downloading github.com/jcelliott/lumber v0.0.0-20160324203708-dd349441af25
go: downloading gopkg.in/src-d/go-billy.v4 v4.3.2
go: downloading gopkg.in/yaml.v2 v2.4.0
go: downloading github.com/pkg/errors v0.8.1
ERROR: ld.so: object '/lib/x86_64-linux-gnu/liboneagentproc.so' from /etc/ld.so.preload cannot be preloaded (cannot open shared object file): ignored.
```
But that's okay, just disable process injection on OneAgent....

**Comment by patrickpichler** [comment]
I guess this was closed by #72

**Comment by UnseenWizzard** [comment]
I'd think so. @baumgarb did my change solve this, or is there anything we're still missing in your opinion?

**Comment by CruzanCaramele** [comment]
closing this as solved by #72
